Source/common/SNTMetricSet.m

```diff
@@ -605,10 +605,15 @@ - (void)addConstantBooleanWithName:(NSString *)name
 
 /** Export current state of the SNTMetricSet as an NSDictionary. */
 - (NSDictionary *)export {
-  NSDictionary *exported = nil;
+  NSDictionary *exported;
+
+  NSArray *callbacks;
+  @synchronized(self) {
+    callbacks = [_callbacks mutableCopy];
+  }
 
   // Invoke callbacks to ensure metrics are up to date.
-  for (void (^cb)(void) in _callbacks) {
+  for (void (^cb)(void) in callbacks) {
     cb();
   }
```

Comment:
Shouldn't this be:

```c++
@synchronized(self) {
  callbacks = [_callbacks mutableCopy];
  
   for (void (^cb)(void) in callbacks) {
    cb();
   }
}
```

Reply:
That would hold a lock while we're calling each callback with no control over what they're doing, so not only could they take quite a while to run they could themselves try to grab locks of some kind. The race seems to be in adding callbacks to the array while we're iterating it. Copying the array avoids this problem as long as the callbacks themselves don't change (which they shouldn't be able to as blocks).